File: app/src/main/java/com/shevaalex/android/rickmortydatabase/RmApplication.java
package com.shevaalex.android.rickmortydatabase;

import android.app.Application;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.os.Build;
import android.os.StrictMode;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.preference.PreferenceManager;

import java.util.Locale;

public class RmApplication extends Application {
    private static final String LIST_PREFERENCE_KEY = "theme_list";
    public static String defSystemLanguage;

    @Override
    public void onCreate() {
        super.onCreate();
        setupTheme();
        //set Strict mode in debug builds
        if (BuildConfig.DEBUG) {
            StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
                    .detectDiskReads()
                    .detectDiskWrites()
                    .detectAll()
                    .penaltyLog()
                    .build());
            StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
                    .detectLeakedSqlLiteObjects()
                    .detectLeakedClosableObjects()
                    .penaltyLog()
                    .penaltyDeath()
                    .build());
        }
        defSystemLanguage = Locale.getDefault().getLanguage();
    }

    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        defSystemLanguage = newConfig.locale.getLanguage();
    }

    //get saved preferences or set default theme if none stored
    public void setupTheme() {
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
        int savedListPrefTheme = Integer.parseInt(sharedPreferences.getString(LIST_PREFERENCE_KEY, "99"));
        boolean savedSwitchThemeNightOn = sharedPreferences.getBoolean("theme_switch", false);
        if (Build.VERSION.SDK_INT >= 29) {
            if (savedListPrefTheme == 99) {
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM);
            } else if (savedListPrefTheme == AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
                    || savedListPrefTheme == AppCompatDelegate.MODE_NIGHT_NO
                    || savedListPrefTheme == AppCompatDelegate.MODE_NIGHT_YES) {
                AppCompatDelegate.setDefaultNightMode(savedListPrefTheme);
            }
        } else {
            int nightMode = AppCompatDelegate.MODE_NIGHT_NO;
            if (savedSwitchThemeNightOn) {
                nightMode = AppCompatDelegate.MODE_NIGHT_YES;
            }
            AppCompatDelegate.setDefaultNightMode(nightMode);
        }
    }
}

File: app/src/main/java/com/shevaalex/android/rickmortydatabase/RmApplication.kt
package com.shevaalex.android.rickmortydatabase

import android.app.Application
import android.os.Build
import android.os.StrictMode
import android.os.StrictMode.VmPolicy
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatDelegate
import androidx.preference.PreferenceManager
import com.shevaalex.android.rickmortydatabase.di.AppComponent
import com.shevaalex.android.rickmortydatabase.di.DaggerAppComponent
import com.shevaalex.android.rickmortydatabase.utils.Constants.Companion.LIST_THEME_PREFERENCE_KEY
import com.shevaalex.android.rickmortydatabase.utils.Constants.Companion.SWITCH_THEME_PREFERENCE_KEY
import com.shevaalex.android.rickmortydatabase.utils.CustomDebugTree
import timber.log.Timber


class RmApplication : Application() {

    val appComponent: AppComponent by lazy {
        DaggerAppComponent.factory().create(this)
    }


    override fun onCreate() {
        super.onCreate()
        setupTimber()
        setupTheme()
        //set Strict mode in debug builds
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            setupStrictMode()
        } else {
            setupStrictModeLegacy()
        }
    }

    private fun setupTimber() {
        if (BuildConfig.DEBUG) {
            Timber.plant(CustomDebugTree())
        }
    }

    //get saved preferences or set default theme if none stored
    private fun setupTheme() {
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
        val savedListPrefTheme =
                sharedPreferences.getString(LIST_THEME_PREFERENCE_KEY, "99")?.toInt()
        val savedSwitchThemeNightOn =
                sharedPreferences.getBoolean(SWITCH_THEME_PREFERENCE_KEY, false)
        if (Build.VERSION.SDK_INT >= 29) {
            if (savedListPrefTheme == 99) {
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
            } else if (savedListPrefTheme == AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM || savedListPrefTheme == AppCompatDelegate.MODE_NIGHT_NO || savedListPrefTheme == AppCompatDelegate.MODE_NIGHT_YES) {
                AppCompatDelegate.setDefaultNightMode(savedListPrefTheme)
            }
        } else {
            when (savedSwitchThemeNightOn) {
                true -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
                false -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
            }
        }
    }

    @RequiresApi(Build.VERSION_CODES.P)
    private fun setupStrictMode() {
        if (BuildConfig.DEBUG) {
            StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.Builder()
                    .detectDiskReads()
                    .detectDiskWrites()
                    .detectAll()
                    .penaltyLog()
                    .build())
            StrictMode.setVmPolicy(VmPolicy.Builder()
                    .detectNonSdkApiUsage()
                    .detectLeakedSqlLiteObjects()
                    .detectLeakedClosableObjects()
                    .penaltyLog()
                    .build())
        }
    }

    private fun setupStrictModeLegacy() {
        if (BuildConfig.DEBUG) {
            StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.Builder()
                    .detectDiskReads()
                    .detectDiskWrites()
                    .detectAll()
                    .penaltyLog()
                    .build())
            StrictMode.setVmPolicy(VmPolicy.Builder()
                    .detectLeakedSqlLiteObjects()
                    .detectLeakedClosableObjects()
                    .penaltyLog()
                    .build())
        }
    }

}